src/coreclr/jit/assertionprop.cpp

```diff
@@ -5053,6 +5054,75 @@ static GCInfo::WriteBarrierForm GetWriteBarrierForm(ValueNumStore* vnStore, Valu
+                {
+                    // No barrier needed if current class is byref-like
+                    // and the destination is an implicit "this"
+                    return comp->eeIsByrefLike(comp->info.compClassHnd) ? GCInfo::WriteBarrierForm::WBF_NoBarrier
+                                                                        : GCInfo::WriteBarrierForm::WBF_BarrierUnknown;
+                }
+
+                // Find the corresponding argument index. Basically, we need to ignore the hidden args
+                // such as generic context, return buffer, etc..
+                unsigned ilArg = comp->compMap2ILvarNum(lclNum);
+                switch (static_cast<int>(ilArg))
+                {
+                    // Ignore non-user args
+                    case ICorDebugInfo::RETBUF_ILNUM:
+                    case ICorDebugInfo::VARARGS_HND_ILNUM:
+                    case ICorDebugInfo::TYPECTXT_ILNUM:
+                    case ICorDebugInfo::UNKNOWN_ILNUM:
+                        return GCInfo::WriteBarrierForm::WBF_BarrierUnknown;
+
+                    default:
+                        break;
+                }
+                assert(ilArg != BAD_VAR_NUM);
+
+                // A small inconsistency: various signature walking VM APIs (e.g. getArgClass)
+                // do not expose implicit "this" args, while compMap2ILvarNum does.
+                if (comp->info.compMethodInfo->args.hasImplicitThis())
+                {
+                    assert(ilArg > 0);
+                    ilArg--;
+                }
+
+                // First, get corresponding CORINFO_ARG_LIST_HANDLE arg data
+                CORINFO_ARG_LIST_HANDLE currentArg = comp->info.compMethodInfo->args.args;
+                for (unsigned i = 0; i < ilArg; i++)
+                {
+                    currentArg = comp->info.compCompHnd->getArgNext(currentArg);
+                }
+
+                // Now get its type and strip `ref` so we can get CORINFO_CLASS_HANDLE of the
+                // underlying type
+                CORINFO_CLASS_HANDLE byrefCls = comp->eeGetArgClass(&comp->info.compMethodInfo->args, currentArg);
+                if (byrefCls != NO_CLASS_HANDLE)
+                {
+                    CORINFO_CLASS_HANDLE actualCls;
+                    CorInfoType          actualClsType = comp->info.compCompHnd->getChildType(byrefCls, &actualCls);
+
+                    // Only structs can be byref-like
+                    if (actualClsType == CORINFO_TYPE_VALUECLASS)
+                    {
+                        assert(actualCls != NO_CLASS_HANDLE);
+                        if (comp->eeIsByrefLike(actualCls))
+                        {
+                            // It's byref-like, barrier is not needed
+                            return GCInfo::WriteBarrierForm::WBF_NoBarrier;
+                        }
+                    }
+                }
+            }
+            return GCInfo::WriteBarrierForm::WBF_BarrierUnknown;
```

Comment:
This look potentially quite expensive. You may want to try running an SPMI collection on this PR to get more accurate diffs, there are a lot of misses.